src/structures/lists/digest.rs

```diff
@@ -1,11 +1,72 @@
 // Copyright 2020 Contributors to the Parsec project.
 // SPDX-License-Identifier: Apache-2.0
+use crate::constants::algorithm::HashingAlgorithm;
 use crate::structures::Digest;
-use crate::tss2_esys::TPML_DIGEST;
+use crate::tss2_esys::{TPML_DIGEST, TPML_DIGEST_VALUES, TPMT_HA, TPMU_HA};
 use crate::{Error, Result, WrapperErrorKind};
 use log::error;
+use std::collections::HashMap;
 use std::convert::{TryFrom, TryInto};
 
+#[derive(Debug, Clone, Default)]
+pub struct DigestValues {
+    digests: HashMap<HashingAlgorithm, Digest>,
+}
+
+impl DigestValues {
+    pub fn new() -> Self {
+        DigestValues {
+            digests: HashMap::new(),
+        }
+    }
+
+    pub fn set(&mut self, alg: HashingAlgorithm, dig: Digest) {
+        let _ = self.digests.insert(alg, dig);
+    }
+}
+
+impl TryFrom<(HashingAlgorithm, Digest)> for TPMT_HA {
+    type Error = Error;
+    fn try_from(digest: (HashingAlgorithm, Digest)) -> Result<Self> {
+        let algid: crate::tss2_esys::TPM2_ALG_ID = digest.0.into();
+        let digest_val = digest.1;
+        Ok(TPMT_HA {
+            hashAlg: algid,
+            digest: match digest.0 {
+                HashingAlgorithm::Sha1 => TPMU_HA {
+                    sha1: digest_val.try_into()?,
+                },
+                HashingAlgorithm::Sha256 => TPMU_HA {
+                    sha256: digest_val.try_into()?,
+                },
+                HashingAlgorithm::Sha384 => TPMU_HA {
+                    sha384: digest_val.try_into()?,
+                },
+                HashingAlgorithm::Sha512 => TPMU_HA {
+                    sha512: digest_val.try_into()?,
+                },
+                HashingAlgorithm::Sm3_256 => TPMU_HA {
+                    sm3_256: digest_val.try_into()?,
+                },
+                _ => return Err(Error::local_error(WrapperErrorKind::UnsupportedParam)),
+            },
+        })
+    }
+}
+
+impl TryFrom<DigestValues> for TPML_DIGEST_VALUES {
+    type Error = Error;
+    fn try_from(digest_values: DigestValues) -> Result<Self> {
+        let mut tss_digest_values: TPML_DIGEST_VALUES = Default::default();
+        for (digest_hash, digest_val) in digest_values.digests.iter() {
+            tss_digest_values.digests[tss_digest_values.count as usize] =
+                (*digest_hash, digest_val.clone()).try_into()?;
```

Comment:
The `clone` here can be avoided since `digest_values` is dropped at the end of this method anyway, you can call `digest_values.digests.drain()` above and get ownership of the hashmap contents.